### PR TITLE
Add new backbones for registration

### PR DESCRIPTION
**Pull request by humanpose1**
Here you can find KPConv and Pointnet for registration.
For KPConv, the training takes more time (In comparison with Minkowski)
Pointnet2 doesn't work yet (Maybe a problem of hyperparams)

### CODE REVIEW

File: conf/data/registration/fragment3dmatch_sparse.yaml [hunk 1]
```diff
@@ -19,6 +19,8 @@ data:
         size: ${data.first_subsampling}
   train_transform:
     - transform: SaveOriginalPosId
+    - transform: FixedPoints
```
Comment (nicolas-chaulet) [inline conf/data/registration/fragment3dmatch_sparse.yaml:22]: Sure about that for sparse?

File: conf/training/kpconv.yaml [hunk 2]
```diff
@@ -18,7 +18,7 @@ training:
         bn_scheduler:
             bn_policy: "step_decay"
             params:
-                bn_momentum: 0.02
+                bn_momentum: 0.98
```
Comment (nicolas-chaulet) [inline conf/training/kpconv.yaml:21]: 0.02 is correct I believe, it is the reverse convention to tensorflow

File: conf/training/kpconv_reg.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            class: SGD
+            params:
+                momentum: 0.98
+                lr: ${training.optim.base_lr} # The path is cut from training
+                weight_decay: 1e-3
+        lr_scheduler: ${lr_scheduler}
+        bn_scheduler:
+            bn_policy: "step_decay"
+            params:
+                bn_momentum: 0.98
```
Comment (nicolas-chaulet) [inline conf/training/kpconv_reg.yaml:21]: 0.02

File: torch_points3d/datasets/registration/general3dmatch.py [hunk 4]
```diff
@@ -9,12 +9,13 @@
...
 from torch_points3d.datasets.registration.utils import PatchExtractor
 from torch_points3d.datasets.registration.utils import tracked_matches
-from torch_points3d.datasets.registration.pair import Pair
+from torch_points3d.datasets.registration.pair import Pair, MultiScalePair
 from torch_points3d.metrics.registration_tracker import PatchRegistrationTracker
 from torch_points3d.metrics.registration_tracker import FragmentRegistrationTracker
 from torch_points3d.core.data_transform.transforms import GridSampling
 from torch_points3d.datasets.registration.base_siamese_dataset import BaseSiameseDataset
 from torch_points3d.datasets.registration.utils import compute_overlap_and_matches, compute_subsampled_matches
+from torch_points_kernels.points_cpu import ball_query
```
Comment (nicolas-chaulet) [inline torch_points3d/datasets/registration/general3dmatch.py:18]: Unused import, please remove

File: torch_points3d/datasets/registration/general3dmatch.py [hunk 5]
```diff
@@ -152,6 +153,8 @@ def get_patch_offline(self, idx):
         if self.transform is not None:
             data_source = self.transform(data_source)
             data_target = self.transform(data_target)
+
+        batch = Pair.make_pair(data_source, data_target)
```
Comment (nicolas-chaulet) [inline torch_points3d/datasets/registration/general3dmatch.py:157]: Duplicate line

File: conf/data/registration/fragment3dmatch_sparse.yaml (hunk 1 above)
Comment (humanpose1) [reply conf/data/registration/fragment3dmatch_sparse.yaml:22]: Pointcloud can be very large(up to 100k points) so I limit the size.

File: conf/data/registration/fragment3dmatch.yaml [hunk 6]
```diff
@@ -10,14 +10,69 @@ data:
...
-  train_transforms:
+  train_transform:
+    - transform: SaveOriginalPosId
     - transform: RandomNoise
       params:
-      sigma: 0.01
-      clip: 0.05
+        sigma: 0.005
+        clip: 0.05
+    - transform: RandomRotate
```
Comment (tchaton) [inline conf/data/registration/fragment3dmatch.yaml:26]: You should use the 3D rotate transform. Should be faster and more compact config file.

File: conf/data/registration/fragment3dmatch.yaml [hunk 7]
```diff
@@ -10,14 +10,69 @@ data:
...
+        axis: 2
+    - transform: RandomScaleAnisotropic
+      params:
+        scales: [0.9,1.2]
+
+    - transform: GridSampling
+      params:
+        size: ${data.first_subsampling}
+        quantize_coords: False
+        mode: "mean"
```
Comment (tchaton) [inline conf/data/registration/fragment3dmatch.yaml:46]: You should use mode: "last" to retain real point position in transforms, and mean in pre_transforms.

File: conf/training/kpconv.yaml (hunk 2 above)
Comment (tchaton) [reply conf/training/kpconv.yaml:21]: Could explain why it is so long to train.

File: conf/training/kpconv_reg.yaml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    cuda: 1
+    precompute_multi_scale: True # Compute multiscate features on cpu for faster training / inference
+    optim:
+        base_lr: 0.01
+        grad_clip: 100
+        optimizer:
+            class: SGD
+            params:
+                momentum: 0.98
+                lr: ${training.optim.base_lr} # The path is cut from training
```
Comment (tchaton) [inline conf/training/kpconv_reg.yaml:15]: # The path is cut from training can be removed

File: torch_points3d/datasets/registration/general3dmatch.py [hunk 9]
```diff
@@ -152,9 +150,12 @@ def get_patch_offline(self, idx):
         if self.transform is not None:
             data_source = self.transform(data_source)
             data_target = self.transform(data_target)
-        batch = Pair.make_pair(data_source, data_target)
-        batch = batch.contiguous().to(torch.float)
-        return batch
+
+        if(hasattr(data_source, "multiscale")):
+            batch = MultiScalePair.make_pair(data_source, data_target)
```
Comment (tchaton) [inline torch_points3d/datasets/registration/general3dmatch.py:155]: MultiScalePair awesome :)

File: torch_points3d/models/registration/kpconv.py [hunk 10]
```diff
@@ -113,3 +122,178 @@ def backward(self):
...
+
+        output = self.FC_layer(data.x)
+        if self.normalize_feature:
+            return output / (torch.norm(output, p=2, dim=1, keepdim=True) + 1e-3)
+        else:
+            return output
+
+    def forward(self) -> Any:
+        """Run forward pass. This will be called by both functions <optimize_parameters> and <test>."""
+        self.output = self.apply_nn(self.input, self.pre_computed, self.upsample)
```
Comment (tchaton) [inline torch_points3d/models/registration/kpconv.py:224]: I usually name apply_nn -> _forward, up to you.

File: torch_points3d/models/registration/kpconv.py [hunk 11]
```diff
@@ -113,3 +122,178 @@ def backward(self):
...
+            hard_pairs = self.miner_module(output, labels)
+        # loss
+        return self.metric_loss_module(output, labels, hard_pairs)
+
+    def compute_loss(self):
+        self.loss = 0
+
+        # Get regularization on weights
+        if self.lambda_reg:
+            self.loss_regul = self.get_regularization_loss(regularizer_type="l2", lambda_reg=self.lambda_reg)
```
Comment (tchaton) [inline torch_points3d/models/registration/kpconv.py:258]: @nicolas-chaulet I think it is also what weight decay does right in the optimiser right ?

File: torch_points3d/models/registration/kpconv.py [hunk 12]
```diff
@@ -113,3 +122,178 @@ def backward(self):
...
+        self.loss += self.loss_reg
+
+    def backward(self):
+        """Calculate losses, gradients, and update network weights; called in every training iteration"""
+        # caculate the intermediate results if necessary; here self.output has been computed during function <forward>
+        # calculate loss given the input and intermediate results
+        if hasattr(self, "loss"):
+            self.loss.backward()  # calculate gradients of network G w.r.t. loss_G
+
+    def get_output(self):
```
Comment (tchaton) [inline torch_points3d/models/registration/kpconv.py:278]: Would be better to always return the same number of objects as it is safer 
```
        if self.match is not None:
             return self.output, self.output_target
         else:
             return self.output, None
```

File: torch_points3d/models/registration/kpconv.py [hunk 13]
```diff
@@ -113,3 +122,178 @@ def backward(self):
...
+        if self.match is not None:
+            return self.output, self.output_target
+        else:
+            return self.output
+
+    def get_batch(self):
```
Comment (tchaton) [inline torch_points3d/models/registration/kpconv.py:284]: same here

File: torch_points3d/models/registration/kpconv.py [hunk 14]
```diff
@@ -113,3 +122,178 @@ def backward(self):
...
+        if self.match is not None:
+            batch = self.input.batch
+            batch_target = self.input_target.batch
+            return batch, batch_target
+        else:
+            return None
+
+    def get_input(self):
```
Comment (tchaton) [inline torch_points3d/models/registration/kpconv.py:292]: same here

File: torch_points3d/models/registration/pointnet2.py [hunk 15]
```diff
@@ -97,8 +100,155 @@ def backward(self):
             self.loss.backward()  # calculate gradients of network G w.r.t. loss_G
 
 
-class FragmentPointNet(UnwrappedUnetBasedModel):
+class FragmentPointNet2_D(UnetBasedModel):
```
Comment (tchaton) [inline torch_points3d/models/registration/pointnet2.py:103]: I think you could have a FragmentBaseModel, as there are lot of duplicated code.

File: torch_points3d/models/registration/kpconv.py (hunk 11 above)
Comment (nicolas-chaulet) [reply torch_points3d/models/registration/kpconv.py:258]: yes

**Comment by tchaton** [comment]
@humanpose1 Looks good to me. Thanks for updating your code. It would be great if you could turn your models within the API.

And It would be nice if you could the API directly to create KPConv, PointNet2, ME (needs to be added). 

I think it is time to make the API grow as much as possible for more people to get involved in the framework.

Best,
T.C
